[te] Fix casting of unsigned char, and abs(int)

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44152 Fix frac in CUDA fuser
* #44160 Prevent the TE fuser from getting datatypes it can't handle
* **#44157 [te] Fix casting of unsigned char, and abs(int)**

Differential Revision: [D23528507](https://our.internmc.facebook.com/intern/diff/D23528507)

bertmaher committed:
[te] Fix casting of unsigned char, and abs(int)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Fix casting of unsigned char, and abs(int)"




[ghstack-poisoned]

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -1152,20 +1152,48 @@ def rand(dtype, device="cuda"):
             torch.int64,
-            torch.float16,
-            torch.bfloat16,
+            # torch.float16,
```
Nit: maybe add a comment whether we want these to be supported at all?

bertmaher replied:
Oh, yeah, float16 should be coming in soon-ish.  The commented lines are a transient state for this stack :-)

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -186,7 +186,7 @@ void CudaPrinter::visit(const For* v) {
 
 void CudaPrinter::visit(const Cast* v) {
-  os() << cudaDtypeCppString(v->dtype());
+  os() << "(" << cudaDtypeCppString(v->dtype()) << ")";
```
Ouch, how did it work? :)

bertmaher replied:
I don't know when it became a thing but apparently you can cast by saying `float(x)`, like it's a constructor.  But that sorta doesn't parse if you do `unsigned char(x)`.  Clearly we need whitespace in identifiers! (I hope this reddit post is a joke: https://www.reddit.com/r/ProgrammerHumor/comments/4on81i/til_c_allows_u200b_zero_width_space_in_identifiers/ )

ZolotukhinM reviewed (approved):
Looks good! I guess you'll need to replace `self.assertEqual(ref, t(x))` with `np.testing.assert_allclose(...)` to appease the CI gods though.

bertmaher committed:
Update on "[te] Fix casting of unsigned char, and abs(int)"




[ghstack-poisoned]

bertmaher committed:
Update on "[te] Fix casting of unsigned char, and abs(int)"


Differential Revision: [D23528507](https://our.internmc.facebook.com/intern/diff/D23528507)

[ghstack-poisoned]

ngimel commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -209,6 +209,9 @@ void CudaPrinter::visit(const Intrinsics* v) {
   if (returnType == ScalarType::Half || returnType == ScalarType::Float) {
     func_name = func_name + "f";
   }
+  if (v->op_type() == IntrinsicsOp::kFabs && is_integral(returnType)) {
```
does nvrtc not handle std::abs or ::abs? All this exp/expf is so last decade (and is not used in eager part of the codebase at all).

bertmaher replied:
I think it does, actually.  I don't know why we went the `+ "f"` route, except that maybe the old fuser uses it?  I'll take a note to use `std::` in a follow-up.

bertmaher committed:
Update on "[te] Fix casting of unsigned char, and abs(int)"


Differential Revision: [D23528507](https://our.internmc.facebook.com/intern/diff/D23528507)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Fix casting of unsigned char, and abs(int)"


Differential Revision: [D23528507](https://our.internmc.facebook.com/intern/diff/D23528507)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Fix casting of unsigned char, and abs(int)"


Differential Revision: [D23528507](https://our.internmc.facebook.com/intern/diff/D23528507)

[ghstack-poisoned]
